File: static/src/static/notifications/telegram.py
from __future__ import annotations

import json
import urllib.request
from typing import TYPE_CHECKING

from .base import NotificationChannel

if TYPE_CHECKING:
    from static.models import Finding


class TelegramChannel(NotificationChannel):
    def __init__(self, token: str, chat_id: str) -> None:
        self.token = token
        self.chat_id = chat_id

    def send(self, summary: str, findings: list[Finding]) -> None:
        if not self.token or not self.chat_id:
            return
        try:
            url = f"https://api.telegram.org/bot{self.token}/sendMessage"
            data = {"chat_id": self.chat_id, "text": summary, "parse_mode": "Markdown"}
            encoded_data = json.dumps(data).encode("utf-8")
            req = urllib.request.Request(
                url,
                data=encoded_data,
                headers={"Content-Type": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=5):
                pass
        except Exception:
            pass

File: dynamic/internal/notify/telegram.go
package notify

import (
	"bytes"
	"encoding/json"
	"net/http"
	"time"

	"github.com/yetanotherparticipant/PipeSec/dynamic/internal/dynscan"
)

type TelegramChannel struct {
	token  string
	chatID string
	client *http.Client
}

func NewTelegramChannel(token, chatID string) *TelegramChannel {
	return &TelegramChannel{
		token:  token,
		chatID: chatID,
		client: &http.Client{Timeout: 5 * time.Second},
	}
}

func (t *TelegramChannel) Send(msg string, findings []dynscan.Finding) {
	if t == nil || t.token == "" || t.chatID == "" {
		return
	}
	body, _ := json.Marshal(map[string]string{
		"chat_id":    t.chatID,
		"text":       msg,
		"parse_mode": "Markdown",
	})
	req, err := http.NewRequest(
		http.MethodPost,
		"https://api.telegram.org/bot"+t.token+"/sendMessage",
		bytes.NewReader(body),
	)
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := t.client.Do(req)
	if err != nil {
		return
	}
	resp.Body.Close()
}
